File: python_code/data analysis/homebrew origin/parser.py
def read_file(filename):
    """
    This reads the data from the given file. If invalid data is given, float("NaN") is used.
    :param filename: the filename of the data file to be read
    :return: a list of the data at the file and the column headers of the file in a tuple (data, col_headers)
    """
    file = open(filename, 'r')
    lines = file.readlines()
    header_line = lines.pop(0)

    line_length = len(header_line)
    column_length = 15
    number_of_columns = line_length // column_length

    row_set = []

    for line in lines:
        line = line.strip("\n")
        line_data = []

        for i in range(number_of_columns):
            index_start = i * column_length
            index_end = index_start + column_length
            substring = line[index_start: index_end]
            if substring == "***************":
                line_data.append(float("NaN"))
            else:
                line_data.append(float(substring))
        row_set.append(line_data)

    col_set = convert_row_set_to_col_set(row_set)

    header_list = []
    for i in range(number_of_columns):
        index_start = i * column_length
        index_end = index_start + column_length
        substring = header_line[index_start: index_end]
        substring = substring.replace(" ", "")
        header_list.append(substring)

    return col_set, header_list


def get_metadata(filename: str):
    """
    Takes a zenus data filename and outputs a dictionary with the day of year and year inside.
    :param filename: filename like this: OH_Andover_ALO23day1.dat or OH_Andover_ALO08day300.dat
    :return: a dict with the key:values  "year" and "day_of_year", where "year" is 20YY.
    """
    metadata = {}
    date = filename.replace("OH_Andover_ALO", "").replace(".dat", "")
    year_stub, day_of_year = tuple(date.split("day"))

    metadata["year"] = "20" + year_stub
    metadata["day_of_year"] = day_of_year
    return metadata


def convert_row_set_to_col_set(dataset):
    row_total = len(dataset)
    col_total = len(dataset[0])
    col_set = []
    for i in range(col_total):
        col_set.append([])

    for i in range(row_total):
        for j in range(col_total):
            col_set[j].append(dataset[i][j])
    return col_set

File: multi_platform_projects/homebrew origin/java_src/Parser.java
import java.util.ArrayList;
import java.io.File;
import java.util.Scanner;
import java.lang.Double;

public class Parser {
    /**
     * This reads the data from the given file. If invalid data is given, null object is used instead of a Double object in that position in the output data.
     *  @param filename the path to and the name of the file that you want read.
     *  @return 2D ArrayList<ArrayList<Double>> where each inner array has a column of data.
     */
    public static ArrayList<ArrayList<Double>> parseFile(String filename) {
        File file = new File(filename);
        ArrayList<String> lines = new ArrayList<String>();

        try (Scanner input = new Scanner(file)) {
            while (input.hasNextLine()) {
                String line = input.nextLine();
                line = line.replace("\n", "").replace("\r", "");
                lines.add(line);
            }
        } catch (java.io.FileNotFoundException e) {
            System.out.println("There was an error with opening the file:");
            System.out.println(e);
            System.exit(1);
        }
        
        String headerLine = lines.remove(0);
	int lineLength = headerLine.length();
	int colLength = 15;
	int numberOfColumns = (int)(lineLength / colLength);
	ArrayList<ArrayList<Double>> rowSet = new ArrayList<ArrayList<Double>>();

	for (String line : lines) {
	    ArrayList<Double> lineData = new ArrayList<Double>();
	    for (int i = 0; i < numberOfColumns; i++) {
	        int indexStart = i * colLength;
		int indexEnd = indexStart + colLength;
		String substring = line.substring(indexStart, indexEnd);
		substring = substring.replace(" ", "");
		if (substring.contains("*")) {
		    lineData.add(null);
		} else if (substring.contains("NaN")) {
		    lineData.add(null);
		} else {
		    lineData.add(Double.valueOf(substring));
		}
            }
	    rowSet.add(lineData);
	}

        ArrayList<ArrayList<Double>> colSet = rowSetToColSet(rowSet);

	return colSet; 
    }
    public static ArrayList<ArrayList<Double>> rowSetToColSet(ArrayList<ArrayList<Double>> rowSet) {
        int rowCount = rowSet.size();
	int colCount = rowSet.get(0).size();
	ArrayList<ArrayList<Double>> colSet = new ArrayList<>();

	for (int j = 0; j < colCount; j++) {
	    colSet.add(new ArrayList<Double>());
	}
	for (int i = 0; i < rowCount; i++) {
	    for (int j = 0; j < colCount; j++) {
	        colSet.get(j).add(rowSet.get(i).get(j));
	    }
	}
	return colSet;
    }
}
